### PR TITLE
Viscoelastic plastic v1

### PR DESCRIPTION
**Pull request by lmoresi**
Placeholder for VEP models, and fixing the notch benchmark so we can test them !

Also fixing the labels which appear in issue #106 - an indexing error when a label has no support on a particular process.

L

### CODE REVIEW

**Comment by lmoresi** [comment]
I've made a few adjustments to the MG configuration because I was having trouble with robustness for the Stokes sinker problem, particularly as we refined the grid. That's looking better, though I make no claims for performance or parallel scaling at this stage. I just want all the examples to be able to run reliably in serial / parallel. 

The underworld-Bat experiment is working nicely in serial and parallel regardless of resolution. It's good to get a reliable solve for this one because it is the case that troubled Spiegelman et al ... though we have not used physical scaling in this model so maybe it's just easier.  I think we should run this with a green and red colormap and make it into a Christmas card

Screen Shot 2022-12-13 at 9 55 13 am

**Comment by bknight1** [comment]
Sounds good to me! Shall I commit the scaling module in this branch so we can include scaling in future examples?

**Comment by lmoresi** [comment]
@bknight1 - if the scaling is passing the tests etc, feel free to add it here. Julian and I can finish this up on Tuesday.

**Comment by bknight1** [comment]
Scaling works well, although had to remove converting swarm and mesh variables directly as I wasn't sure how to add access to the variables.

**Comment by lmoresi** [comment]
Yep, tests passing again.

File: underworld3/scaling/_scaling.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+              mass**(dmass) *
+              temperature**(dtemp) *
+              substance**(dsubstance))
+
+    ''' This section needs to be modified for UW3 swarm/mesh vars '''
+
+    if (isinstance(value, uw.discretisation._MeshVariable) or
+       isinstance(value, uw.swarm.SwarmVariable)):
+
+        print('swarm/mesh objects not currently supported')
```
Comment (bknight1) [inline underworld3/scaling/_scaling.py:185]: Need to add write access to mesh/swarm variables for conversion (if we want)

**Comment by lmoresi** [comment]
Adding some functionality to support refinement and adaptivity (and something similar: reading information saved in a mesh at different resolution).

  - added the ability for a swarm to be transferred to a different mesh (automatically rebuilding the proxy variables)
  - added rbf_interpolation to mesh variables to map between swarms and multiple meshes (k-d tree based).
  - added mesh adaptation using a scalar field (from meshVar) to drive refinement / coarsening
  - fixed the mesh integral capability (a small bug in our wrapper code).
  - also adding a mesh save / read that is rbf / swarm based so that mesh permutation or (modest) changes in mesh connectivity or mesh resolution can be handled.

Overall, the code needs quite a bit of refactoring before we release it, but we can keep benchmarking and writing new tests while we chop stuff out. 

This is a demo: *mesh is adapted using the y coordinate as a refinement indicator. The colour-field is a swarm variable that was defined using the original uniform-resolution mesh*

Screen Shot 2023-02-16 at 8 52 14 pm

**Comment by lmoresi** [comment]
merging in Ben / julian's changes and adding the bits and pieces that we used to fix up the adaptivity with swarms (at least fixed up wrt tests ... )

  - A swarm migration routine that passes around particles that would otherwise be lost when the mesh is re-partitioned during adaptation (this is a band-aid, there is certainly a better, scalable way to do this ... and maybe @knepley already has this on a branch somewhere, or maybe the adaptation is not correct)
  - Some swarm pop-control that works by periodically resetting the swarm (so it becomes a streak). This saves a lot of thinking about where to put new points etc, but may cause trouble if the streaks are truncated too quickly
  - added a very simple, general (swarm-to-swarm) rbf interpolation routine using kdtree and add it to swarm vars and mesh vars to go back and forth (collecting up various half-implementations around the place).

**Comment by lmoresi** [comment]
A word about tests ... I noticed that pytest is relying on our garbage collection working perfectly across all the mesh / solver instantiation. We eventually find things are blowing up - including with the options database and all sorts of places that we don't really expect to break. All works if we run only the failed tests. 

So at least for these checks, I suggest that we run tests in batches (changing yaml accordingly)

**Comment by lmoresi** [comment]
Can we set up some convergence tests for the viscoplastic cases. I found that the Jacobians for cases with nested sympy.Max / sympy.Min operators sometimes completely fail to converge and I could not tell quite why or when it would happen. Comparing with constants does not seem too bad though.

As a principle, perhaps we should also take care to write our code to be sure it simplifies correctly ... that is, sympy can see through to make viscoplastic into a pure viscous rheology if no yield is set. That's the benefit of using sympy !

**Comment by bknight1** [comment]
Hmmm, I have found that models struggle to converge when trying to implement [yielding](https://github.com/underworld-community/UW3-benchmarks/blob/9d747e0857d1c72131c5425a2bec61763d90f1ba/Working/Cartesian/Ex_stokes_notch.py), but models converge fine with a [SR dependent viscosity](https://github.com/underworld-community/UW3-benchmarks/blob/dev/Working/Cartesian/Ex_stokes_slabDetachment.py), where both have a viscosity limiter.

**Comment by knepley** [comment]
Feel free to file this away, but I have been wanting to do it for a while.

From the work that Dave May and Marc Spiegelman did on visco-plastic (and also Boris Kaus and Toby Keller), it seems like the problem is that Newton steps right beyond the plastic envelope, and then spends a lot of time getting back to it. Dave fixed this by using the fast Picard iterations to get back to the envelope, and Toby fixed it by adding some elasticity to restrict the step. Of course, I have my own fix for both.

I think you might be able to use nonlinear preconditioning of Newton with Richardson to fix the step length problem. Dave and I could never get a code to behave exactly right for that, but now I think we could do it.

For the elasticity idea, the problem is that adding it pollutes the solution. Toby and Boris minimized the pollution, but it is still there. I think the right thing to do here is "precondition" the timestep, namely use the step with elasticity as the initial guess, and then solve the original visco-plastic one. I think this is the best option, but the most complicated.

**Comment by bknight1** [comment]
Thanks for that @knepley. Re-reading over the Spiegelman et al., (2016) paper they do also state that 'Initial numerical experiments suggests that convergence is even more difficult using the minimum viscosity model'...

**Comment by lmoresi** [comment]
> In response to this - I was struggling to get sympy to compute good Jacobians when using Max(Min( )) combinations but it may be something to do with the caching issue that I raised #114. For the Spiegelman et al benchmark, I found that a physically weaker non-linearity (e.g. removing the pressure dependence) works well to get close to the solution but I also agree that things would be simpler if we could simply build them into the math side of the solve.

**Pull request merged**